src/iter_both.rs

```diff
@@ -0,0 +1,133 @@
+use fastq_pair::{delete_empty_fastq, parse_header, parse_read, PartialRead};
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{BufReader, BufWriter, Result};
+use std::io::Write;
+use std::path::Path;
+
+/// Contains all Read/Write objects
+struct IO {
+    in_read1: BufReader<File>,
+    in_read2: BufReader<File>,
+    out_read1: BufWriter<File>,
+    out_read2: BufWriter<File>,
+    out_single: BufWriter<File>,
+    singleton_path: String,
+}
+
+/// Create all IO objects for reading and writing
+fn create_io(r1_path: &str, r2_path: &str) -> Result<IO> {
+    let parent = Path::new(r1_path).parent().unwrap();
+    let r1_out_path = parent.join("R1_paired.fastq").to_str().unwrap().to_string();
+    let r2_out_path = parent.join("R2_paired.fastq").to_str().unwrap().to_string();
+    let singleton_path = parent.join("Singletons.fastq").to_str().unwrap().to_string();
+    // Readers
+    let r1_handle = File::open(&r1_path)?;
+    let r2_handle = File::open(&r2_path)?;
+    let r1_reader = BufReader::new(r1_handle);
+    let r2_reader = BufReader::new(r2_handle);
+    // Writers
+    let r1_out_handle = File::create(&r1_out_path)?;
+    let r2_out_handle = File::create(&r2_out_path)?;
+    let singleton_handle = File::create(&singleton_path)?;
+    let r1_writer = BufWriter::new(r1_out_handle);
+    let r2_writer = BufWriter::new(r2_out_handle);
+    let singleton_writer = BufWriter::new(singleton_handle);
+    Ok(IO {
+        in_read1: r1_reader,
+        in_read2: r2_reader,
+        out_read1: r1_writer,
+        out_read2: r2_writer,
+        out_single: singleton_writer,
+        singleton_path: singleton_path.to_string(),
+    })
+}
+
+
+/// Writes out paired reads two FASTQ files
+fn write_read(header: &str,
+              w1: &mut BufWriter<File>,
+              w2: &mut BufWriter<File>,
+              map1: &mut HashMap<String, PartialRead>,
+              map2: &mut HashMap<String, PartialRead>) -> Result<()> {
+    let r1 = &map1[header];
```

Comment:
Small nitpick here: `let r1 = map1.remove(header)`, and similarly for `r2`, may be slightly more concise.

Reply:
That's much better, didn't even think of checking if `remove` returned something. Thanks!

Reply:
Yeah, I just realized it returned an `Option<T>` rather than `T`, but you should be able to `unwrap` it (it panics, but only where `map1[header]` would panic as well).

Reply:
On an aside, `remove` returns an `Option` which means if I use the `?` operator I get:
```
the trait `std::convert::From<std::option::NoneError>` is not implemented for `std::io::Error` 
```
In this case I'm lame and am planning on adding an `.expect` since this function should never be called unless the header is in both hashmaps, but if I wanted to use `?` I'm not sure how I'd go about it.

Reply:
Yeah, where the error management gets to be sort of a pain, honestly, is converting from `Option` and/or different `Error` types to your returned `Error` type -- this is very tedious to do with just the current stdlib. You can create your own custom `Error` types which implement `From<NoneError>`, but that isn't really worth it unless you're creating a library. Right now the best way that I've found to handle something like this is (using `failure`) `map1.remove(header).ok_or(format_err!("Header {} not found in map", header))?`. Which gets you a decent error message without requiring quite so much boilerplate. I might make a PR switching us over from using `io::Error` everywhere to `failure` for this reason.

Reply:
Ah ok, makes sense. 

> Right now the best way that I've found to handle something like this is (using failure) `map1.remove(header).ok_or(format_err!("Header {} not found in map", header))?`. Which gets you a decent error message without requiring quite so much boilerplate. I might make a PR switching us over from using `io::Error` everywhere to failure for this reason.

That sounds great. I'll look over the `failure` crate when I get a chance. Thanks for the PR review btw!